### PR TITLE
Replace ASSERTV macro with compiler annotation

### PR DESCRIPTION
**Pull request by mattmacy**
Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by lundman** [comment]
Looks like I can use 
```
#define		__debug __pragma(warning(suppress: 4100 4101))
```
on Windows, so at least there is a way.

File: module/zfs/dsl_dataset.c [hunk 1]
```diff
@@ -2843,7 +2843,7 @@ dsl_dataset_stats(dsl_dataset_t *ds, nvlist_t *nv)
 void
 dsl_dataset_fast_stat(dsl_dataset_t *ds, dmu_objset_stats_t *stat)
 {
-	ASSERTV(dsl_pool_t *dp = ds->ds_dir->dd_pool);
+	dsl_pool_t *dp = ds->ds_dir->dd_pool;
```
Comment (behlendorf) [inline module/zfs/dsl_dataset.c]: `__debug` needed.

File: module/zfs/txg.c [hunk 2]
```diff
@@ -813,7 +813,7 @@ txg_sync_waiting(dsl_pool_t *dp)
 void
 txg_verify(spa_t *spa, uint64_t txg)
 {
-	ASSERTV(dsl_pool_t *dp = spa_get_dsl(spa));
+	dsl_pool_t *dp = spa_get_dsl(spa);
```
Comment (behlendorf) [inline module/zfs/txg.c]: `__debug` needed.

File: lib/libspl/include/sys/debug.h [hunk 3]
```diff
@@ -33,4 +33,10 @@
 #define	__printflike(x, y) __attribute__((__format__(__printf__, x, y)))
 #endif
 
+#ifdef ZFS_DEBUG
+#define	__debug
+#else
+#define	__debug __attribute__((__unused__))
+#endif /* ZFS_DEBUG */
```
Comment (behlendorf) [inline lib/libspl/include/sys/debug.h]: Here too.

File: include/os/linux/spl/sys/debug.h [hunk 4]
```diff
@@ -52,6 +52,12 @@
  */
 #define	__printflike(a, b)	__printf(a, b)
 
+#ifdef ZFS_DEBUG
+#define	__debug
+#else
+#define	__debug __attribute__((__unused__))
```
[the remark this answers is not in the document]
Comment (mattmacy) [reply include/os/linux/spl/sys/debug.h]: Nope. The conditional was only added after gcc failed to build. Clang is of course fine with unconditional define

**Pull request merged**